### PR TITLE
Replace mimemagic with marcel

### PR DESCRIPTION
**Pull request by ghiculescu**
Resolves https://github.com/kreeti/kt-paperclip/issues/53

See also https://github.com/rails/marcel/pull/30, https://weblog.rubyonrails.org/2021/3/26/marcel-upgrade-releases/,  https://github.com/carrierwaveuploader/carrierwave/pull/2551

### Known breaking changes

- Content type validators for MS office formats may need to be updated. See https://github.com/rails/marcel/blob/48644abd771ccaa25cdaad53c0ec02beb9ca73c9/lib/marcel/tables.rb for a list of supported types. For example, `validates_attachment :foo, content_type: { content_type: ["application/vnd.ms-excel"] }` may need to be changed to `validates_attachment :foo, content_type: { content_type: ["application/vnd.ms-excel", "application/x-tika-msoffice"] }`.
- Google sheets exports of office formats may not be recognized without additional configuration. Ensure you're using the latest version of `marcel` which includes https://github.com/rails/marcel/pull/36

### CODE REVIEW

**Comment by pboling** [comment]
`mimemagic` may have broader support for filetypes than marcel.  The 0.3.7 release of mimemagic fixed out deploys with kt-paperclip.

This was our commit message, which shows how simple it is to upgrade for MacOS and Linux users:
```
fix: Update mimemagic gem to MIT-licensed release v0.3.7

• Depend on pre-installed shared-mime-info (GPLv2+)
• Update nokogori 1.11.1 => 1.11.2 (forced)

LOCAL DEV HAS NEW DEPENDENCY

brew install shared-mime-info
bundle install

DEPLOYED ENVS HAVE NEW DEPENDENCY

apt-get install shared-mime-info
bundle install
```

**Comment by ghiculescu** [comment]
If this repo is sticking with `mimemagic` (particularly if Rails isn't), I think some instructions on how to install it and `shared-mine-info` would be helpful. My preference is to replace it with something that has no extra dependencies.

**Comment by EthanZeigler** [comment]
Sadly for people using containerized builds or cloudfoundry systems, adding another system dependency can be a real pain in the butt, especially for larger organizations with lots of red tape trying to maintain legacy software

**Comment by pboling** [comment]
Would it be possible / reasonable to make it user choice on which library to use?

**Comment by ghiculescu** [comment]
I would be open to it. Out of curiosity, are you using this gem with Rails or independently? Rails just made the same transition https://weblog.rubyonrails.org/2021/3/26/marcel-upgrade-releases/

**Comment by pboling** [comment]
Our use case is in a Grape (not Rails) app that does load some of the Rails gems, like ActiveRecord, ActiveSupport and ActionMailer.

**Comment by ghiculescu** [comment]
Gotcha. I think it would be possible, but would rather just keep a single dependency in this PR to keep it simple. Could you please test your app against the PR as it sits now? If it causes issues I can fix them before it's merged :)

**Comment by pboling** [comment]
Yep, I will test it soon.

**Comment by EthanZeigler** [comment]
Are there any updates on getting this merged? This is the last blocker for my team to release a fix for all this.

**Comment by pboling** [comment]
Have you tested this branch in your app to see if it works?  They were waiting on feedback from an integrated environment.  I was planning to test it yesterday but had to manage a different hotfix.

**Comment by pboling** [comment]
@ghiculescu is the build failure concerning?

**Comment by EthanZeigler** [comment]
I'll be able to get my team to test it tomorrow.

**Comment by ghiculescu** [comment]
> @ghiculescu is the build failure concerning?

On the one hand, it's the same failure [on master](https://travis-ci.com/github/kreeti/kt-paperclip/builds/220262173), so I don't think this PR is causing it. 

On the other hand, it's a red build, so yes.

**Comment by ghiculescu** [comment]
I've added a list of known breaking changes we came across while testing this internally. Please let me know if there's others to add to the PR + changelog.

**Comment by geoffharcourt** [comment]
@ghiculescu thanks for opening this PR and for beginning the list of potentially breaking changes. FWIW, office documents were broken using the `shared-mime-info` database alone (being identified as ZIP files, probably due to the underlying structure of MS Office files as zipped XML bundles), so they are likely already broken for anyone using one of the post-yank versions of mimemagic.

**Comment by berniechiu** [comment]
we're also eager to have this merged~

File: paperclip.gemspec
Comment (borama) [inline paperclip.gemspec:27]: I think that the `mime-types` dependency is still needed due to [this require in `lib/paperclip.rb`](https://github.com/kreeti/kt-paperclip/pull/54/files#diff-31e78c641c3ccf9fd8438aeb0d530fb7b9dc400878d3e77973948d4f175d4c94R64). At least I wasn't able to start rails with this patch without the gem installed.

Comment (ghiculescu) [reply paperclip.gemspec:27]: Hmmmm, you might be right. It must have worked for me because I have other deps that require `mime-types`.

I don't understand how the tests are passing without this. Any ideas?

Comment (borama) [reply paperclip.gemspec:27]: Thanks, it looks that the `mime-types` gem is indirectly required through cucumber. The Travis log shows that the `mime-types` gem is [indeed installed](https://travis-ci.com/github/kreeti/kt-paperclip/jobs/494334560#L446) and cucumber is [one of its dependants](https://rubygems.org/gems/mime-types/reverse_dependencies?rdeps_query=cucumber).

Comment (ghiculescu) [reply paperclip.gemspec:27]: Nice pickup. Makes sense why it works in CI but would break in the field. Can you please check if https://github.com/kreeti/kt-paperclip/pull/54/commits/4347fab325033078abe26e68af03fd65d971054e fixes things for you?

**Comment by ghiculescu** [comment]
@ssinghi @sbhawsingka I'm not sure who the maintainers are for this gem; I found you both via https://github.com/kreeti/kt-paperclip/graphs/contributors?from=2021-01-01&to=2021-03-29&type=c

Is there anything I can do to help get this PR merged? I realise I only opened it a few days ago; for me (and possibly many others) it's the only thing holding us back from removing `mimemagic` as a dependency.

For anyone else subscribed to this: please test this fork in your apps - at least in CI - and let me know if it works, or if you come across any issues worth documenting in the PR body. Here's a gemfile line to test it:

```ruby
gem "kt-paperclip", git: "https://github.com/ghiculescu/kt-paperclip.git", branch: "use-marcel" # pending https://github.com/kreeti/kt-paperclip/pull/54
```

File: paperclip.gemspec
Comment (borama) [reply paperclip.gemspec:27]: It does! Thank you Alex.

**Comment by borama** [comment]
I can confirm that we run this patch successfully in production (on newest Rails 6.1.3.1). For us this is definitely the better solution than just updating the `mimemagic` gem and having to manage the mime database file as an [external dependency from OS](https://github.com/mimemagicrb/mimemagic#dependencies). Anything else I can help with perhaps?

**Comment by rcsAlex** [comment]
I can confirm that we had no issues running this branch in our QA environment on Rails 6.0.3.6. We're working on upgrading one of our apps to Rails 5.2.5 right now, and I'll report back if there are any issues there.

**Comment by markwilliamson503** [comment]
I can also confirm this works in our QA environment running Rails 6.0.3.6.

**Comment by rocket-turtle** [comment]
It seems that Excels from google sheets are not correctly recognised
Maybe wait for that fix:
* https://github.com/rails/marcel/issues/35
* https://github.com/rails/marcel/pull/36/files

Or make it a major version change.

**Comment by ghiculescu** [comment]
Thanks @rocket-turtle - I added it to the PR body.

**Comment by Liboul** [comment]
@ghiculescu

> For anyone else subscribed to this: please test this fork in your apps - at least in CI - and let me know if it works

Our CI passed. Tested in a production environment, and all 5 of our uses for Paperclip seem to keep working fine 👌 

Adding the dependency to `shared-mime-info` is a no go for us, not only because it is an additional external dependency, environment dependent, but also because was increasing the time to setup the project on macOS by 30 minutes. So we will be using this branch until it is merged or another solution is released. Thanks!

**Comment by natematykiewicz** [comment]
https://github.com/rails/marcel/pull/36 was merged yesterday. Be on the lookout for a version bump.

**Comment by EmmaB** [comment]
Passes CI for us (not tested in production though) saving image files and .xlsx (application/vnd.openxmlformats-officedocument.spreadsheetml.sheet) files to a bucket on S3, on Rails 6.1.3.1. Many thanks for the work!

**Comment by ghiculescu** [comment]
> [rails/marcel#36](https://github.com/rails/marcel/pull/36) was merged yesterday. Be on the lookout for a version bump.

Great. @rocket-turtle wanna see if this fixes your issues?

**Comment by johnnyshields** [comment]
At TableCheck we have also used this branch for several days with no issues processing 1,000+ uploads each day.

How can we get this PR merged and released?

**Comment by rocket-turtle** [comment]
@ghiculescu looks good.

**Comment by cajuncanuck** [comment]
I've also been using this in an application handling various filetypes and haven't come across any issues.

**Comment by EmmaB** [comment]
FYI this fix to marcel 1.0.0 is now blocking bundles with the latest Rails which uses marcel 0.3.3...

**Comment by johnnyshields** [comment]
Guys lets get this one to the finish line... 🏁

**Comment by pankajbatra** [comment]
eagerly waiting for the merge ⏳

**Comment by johnnyshields** [comment]
@ssinghi can you handle this merge? It's a critical one. I've been running this in production now for over 3 weeks.

File: paperclip.gemspec
Comment (CaptainAwesomeDi) [inline paperclip.gemspec:28]:
```suggestion
  s.add_dependency("marcel", "~> 1.0.1")
```

**Review by CaptainAwesomeDi — COMMENTED** [review]
we can bump this to 1.0.1 now

**Comment by marcelcarousel** [comment]
Eager to have this merged

**Comment by agieche** [comment]
Running this on production (Rails 6.1.3.1) for several weeks now without issues. Mainly handling sound files with substantial file reprocessing. Thanks for your commitment 👍

**Comment by kandayo** [comment]
Also running this in production (on multiple Rails services, from `v5.2.5` to `v6.1.3.1`) for over 3 weeks without issues. I haven't come across any issues.

**Comment by johnnyshields** [comment]
@adrianmann please merge this

**Comment by adrianmann** [comment]
@johnnyshields Merge what?

**Comment by johnnyshields** [comment]
@ssinghi ?

**Comment by mark-brooks-roostify** [comment]
Will someone be merging this?  This is a pretty important update.

**Comment by ghiculescu** [comment]
> For anyone else subscribed to this: please test this fork in your apps - at least in CI - and let me know if it works, or if you come across any issues worth documenting in the PR body. Here's a gemfile line to test it:
> 
> ```ruby
> gem "kt-paperclip", git: "https://github.com/ghiculescu/kt-paperclip.git", branch: "use-marcel" # pending https://github.com/kreeti/kt-paperclip/pull/54
> ```

Bumping [this](https://github.com/kreeti/kt-paperclip/pull/54#issuecomment-809529687) for anyone who missed it. We have been running this branch in production for a few months, it's 👍 .

I think the test failures are intermittent.

**Comment by edward-ellis** [comment]
We have been running with no issues since April 4, 2021.  -  [https://buildingenergyscore.energy.gov](https://buildingenergyscore.energy.gov)

**Comment by geoffharcourt** [comment]
Same here, zero issues since early April 2020.

**Comment by ssinghi** [comment]
Thank you for this PR, we'll get it reviewed soon and work on it.

**Review by sbhawsingka — APPROVED** [review]

**Pull request merged**
